scripts/staging/fm.dml

```diff
@@ -0,0 +1,88 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/*
+ * Factorization Machines.
+ */
+
+forward = function(matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] out) {
+  /*
+   * Computes the model. 
+   *
+   * Reference: 
+   *  - Factorization Machines, Steffen Rendle.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - w0: the global bias, of shape (1,).
+   *  - W : the strength of each feature, of shape (d, 1).
+   *  - V : factorized interaction terms, of shape (d, k).
+   *  
+   * Outputs:
+   *  - out : target vector, of shape (n, 1)
+   */
+   
+   out = w0 + (X != 0) %*% W + 0.5 * rowSums( ((X != 0) %*% V)^2 - ((X != 0)^2 %*% V^2) )
+}
+   
+backward = function(matrix[double] dout, matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] dX, matrix[double] dw0, matrix[double] dW, matrix[double] dV) {
+   /*
+    * This function accepts the upstream gradients w.r.t output target vector, and 
+    * returns the gradients of the loss w.r.t the parameters
+    *
+    * Inputs:
+    *  - dout : the gradient of the loss function w.r.t y, of shape (n, 1).
+    *  - X, w0, W, V are as mentioned in the above forward function.
+    *
+    * Outputs:
+    *  - dX : the gradient of loss function w.r.t  X, of shape (n, d).
+    *  - dw0: the gradient of loss function w.r.t w0, of shape (1,).
+    *  - dW : the gradient of loss function w.r.t  W, of shape (d, 1). 
+    *  - dV : the gradient of loss function w.r.t  V, of shape (d, k).
+    */
+    
```

Comment:
Mike, I am not sure about `*` operations. Please correct me here. Thanks :)

Reply:
Best thing to do here would be to create a "gradient check" test in which you test if the gradients computed in this backward pass are correct w.r.t. the forward function.  All of the existing functions have this type of test, and the [`affine test`](https://github.com/apache/systemml/blob/master/scripts/nn/test/grad_check.dml#L57) would be a great template to use.  Basically, in that `grad_check.dml` file, create a new function `fm` similar to the `affine` function, and test each of `X`, `w0`, `W`, `V`, just like the `X`, `W`, and `b` are tested for the affine layer.  To run, use the single-line in the [docs](https://github.com/apache/systemml/tree/master/scripts/nn/test#execution), running from the base of the `scripts` folder (which is where the `nn` folder is located).